packages/desktop-client/src/components/reports/CategorySelector.tsx

```diff
@@ -5,42 +5,87 @@ import {
   type CategoryGroupEntity,
 } from 'loot-core/src/types/models';
 
+import { CheckAll, UncheckAll } from '../../icons/v2';
 import ViewHide from '../../icons/v2/ViewHide';
 import ViewShow from '../../icons/v2/ViewShow';
 import { type CategoryListProps } from '../autocomplete/CategoryAutocomplete';
 import Button from '../common/Button';
+import Text from '../common/Text';
+import View from '../common/View';
 import { Checkbox } from '../forms';
 
+import GraphButton from './GraphButton';
+
 type CategorySelectorProps = {
   categoryGroups: Array<CategoryGroupEntity>;
+  categories: Array<CategoryEntity>;
   selectedCategories: CategoryListProps['items'];
   setSelectedCategories: (selectedCategories: CategoryEntity[]) => null;
 };
 
 export default function CategorySelector({
   categoryGroups,
+  categories,
   selectedCategories,
   setSelectedCategories,
 }: CategorySelectorProps) {
   const [uncheckedHidden, setUncheckedHidden] = useState(false);
 
+  const allCategoriesSelected = categories.every(category =>
+    selectedCategories.map(selected => selected.id).includes(category.id),
+  );
+
+  const allCategoriesUnselected = !categories.some(category =>
+    selectedCategories.map(selected => selected.id).includes(category.id),
+  );
+
+  let selectAll: CategoryEntity[] = [];
+  categoryGroups.map(categoryGroup =>
+    categoryGroup.categories.map(category => selectAll.push(category)),
+  );
```

Comment:
Fundamentally the same thing, but a bit better performant because it has less way loops to perform.

```suggestion
  const selectedCategoryMap = useMemo(() => selectedCategories.map(selected => selected.id), [selectedCategories]);

  const allCategoriesSelected = categories.every(category =>
    selectedCategoryMap.includes(category.id),
  );

  const allCategoriesUnselected = !categories.some(category =>
    selectedCategoryMap.includes(category.id),
  );

  let selectAll: CategoryEntity[] = [];
  categoryGroups.map(categoryGroup =>
    categoryGroup.categories.map(category => selectAll.push(category)),
  );
```

Reply:
Thanks